kong/dns/client.lua

```diff
@@ -0,0 +1,783 @@
+  if answers then -- domain name is IP literal
+    answers.ttl = LONG_LASTING_TTL
+    answers.expire = now() + answers.ttl
+    return answers, nil, tries
+  end
+
+  -- TODO: For better performance, it may be necessary to rewrite it as an
+  --       iterative function.
+  local types = get_search_types(self, name, typ)
+  local names = search_names(name, self.resolv, self.hosts)
+
+  local err
+  for _, qtype in ipairs(types) do
+    for _, qname in ipairs(names) do
+      answers, err = resolve_name_type(self, qname, qtype, cache_only,
+                                       short_key, tries, resolved_names)
+      -- severe error occurred
+      if not answers then
+        return nil, err, tries
+      end
+
+      if not answers.errcode then
+        insert_last_type(self.cache, qname, qtype) -- cache TYPE_LAST
+        return answers, nil, tries
+      end
+    end
+  end
+
+  -- not found in the search iteration
+  return nil, err, tries
+end
+
+
+local function resolve_all(self, name, qtype, cache_only, tries, resolved_names)
+  name = string_lower(name)
+  tries = setmetatable(tries or {}, TRIES_MT)
+
+  -- key like "short:example.com:all" or "short:example.com:5"
+  local key = "short:" .. name .. ":" .. (qtype or "all")
+
+  stats_init_name(self.stats, name)
+  stats_increment(self.stats, name, "runs")
+
+  if detect_recursion(resolved_names, key) then
+    stats_increment(self.stats, name, "fail_recur")
+    return nil, "recursion detected for name: " .. name
+  end
+
+  -- quickly lookup with the key "short:<name>:all" or "short:<name>:<qtype>"
+  local answers, err, hit_level = self.cache:get(key)
+  if not answers then
+    log(DEBUG, "quickly cache lookup ", key, " ans:- hlvl:", hit_level or "-")
+
+    answers, err, tries = resolve_names_and_types(self, name, qtype, cache_only,
+                                                  key, tries, resolved_names)
+    if not cache_only and answers then
+      -- If another worker resolved the name between these two `:get`, it can
+      -- work as expected and will not introduce a race condition.
+
+      -- insert via the `:get` callback to prevent inter-process communication
+      self.cache:get(key, nil, function()
+        return answers, nil, answers.ttl
+      end)
+    end
+
+    stats_increment(self.stats, name, answers and "miss" or "fail")
+
+  else
+    log(DEBUG, "quickly cache lookup ", key, " ans:", #answers, " hlv:", hit_level or "-")
+
+    local ctx = ngx.ctx
+    if ctx and ctx.has_timing then
+      req_dyn_hook_run_hooks("timing", "dns:cache_lookup",
+                             (hit_level and hit_level < HIT_L3))
+    end
+
+    stats_increment(self.stats, name, HIT_LEVEL_TO_NAME[hit_level])
+  end
+
+  -- dereference CNAME
+  if qtype ~= TYPE_CNAME and answers and answers[1].type == TYPE_CNAME then
+    stats_increment(self.stats, name, "cname")
+    return resolve_all(self, answers[1].cname, qtype, cache_only, tries, resolved_names)
```

Comment:
```suggestion
    return resolve_all(self, answers[1].cname, qtype, cache_only, tries, resolved_names, has_timing)
```

Reply:
done https://github.com/Kong/kong/pull/12305/commits/c7fabcddf82eb2aca2e2710e9ba2a2080edbe8d3,
implement a new function _resolve_address to avoid checking for `ngx.ctx.has_timing` in recursion